### ISSUE TITLE
Booting Linux

### ISSUE THREAD

**Issue by wangliwei-intel** [agra-uni-bremen/riscv-vp#21]
Is it capable of booting linux right now? thanks.

**Comment by nmeum** [comment]
Yes, we successfully booted Linux using `linux-vp` (which models a [HiFive Unleashed](https://www.sifive.com/boards/hifive-unleashed)) in the past:

https://user-images.githubusercontent.com/2326560/156125883-c4534d1f-3fd1-450a-a340-197f7fd97c74.mp4

However, we experiment more with embedded operating systems like [Zephyr](https://zephyrproject.org/) and [RIOT](https://www.riot-os.org/) in our own work at the moment. For this reason, depending on your Linux image and your Kernel config, you might run into some minor issues here and there when attempting to boot Linux using `linux-vp`.

If you have any specific question regarding `linux-vp` let us know.

**Comment by Cirromulus** [comment]
If this answered your question, please close this Issue.

**Comment by wangliwei-intel** [comment]
Does it support multicore? thanks.

**Comment by wangliwei-intel** [comment]
@Cirromulus 
Sorry, I reopen this issue.
The riscv-toolchain described in README is a Newlib cross-compiler, so I think that process is for
bare-metal or RTOS mode. 
As you said, I'm interested in linux-vp, Is there any document showing me how to do that?

B. R.
Liwei

**Comment by Cirromulus** [comment]
Yes, the toolchain mentioned in the Readme is for embedded / bare-metal applications (FE320).
The `linux-vp` is compatible to the Sifive Unmatched (FU540), so if you follow tutorials to building linux for that, this will help.
You will have to build an ELF file with integrated rootfs which you can boot. You can then use this file together with the DTB to run on the `linux-vp`.
@nmeum might have a more detailed build process.

**Comment by wangliwei-intel** [comment]
@Cirromulus Thank you.
The system bus is TLM bus, is it possible to replace it with CHI or AXI model?

**Comment by Cirromulus** [comment]
I get the impression that you did not understand what TLM stands for.
This VP is a loosely-timed SystemC TLM model. The actual bus protocol is irrelevant.

**Comment by wangliwei-intel** [comment]
Hi, @nmeum 
I followed [https://github.com/sifive/freedom-u-sdk/tree/2020.08](url) to build the kernel image and rootfs for U540.
Everything looks good, I end up with those files:
![image](https://user-images.githubusercontent.com/13142682/158292035-dd016225-cb6f-4899-8abd-d040f7ec719c.png)

Then I ran the command:
linux-vp --use-data-dmi --tlm-global-quantum=1000 --dtb-file=hifive-unleashed-a00-microsemi.dtb fw_payload.elf

I got this error:
linux-vp: /MyHostHome/Desktop/DockerHome/SystemC-HOME/riscv-vp-rv64/vp/src/platform/common/memory.h:49: void SimpleMemory::write_data(unsigned int, const uint8_t*, unsigned int): Assertion `addr + num_bytes <= size' failed.
Aborted (core dumped)

I think I need to make an ELF image specific for linux-vp, Could you please tell me how?

B. R.
Liwei

**Comment by nmeum** [comment]
I actually used [buildroot](https://buildroot.org) instead of the Freedom-U-SDK in the past. Attached is an older image I generated a few years ago which should boot with current linux-vp (see README.md in the attached tarball). As I mentioned in the earlier comment, we are currently not actively working with Linux, hence you might run into bugs with newer Linux versions or different configurations. If your employer is interested in improving linux-vp feel free to reach out to us via email at `riscv@informatik.uni-bremen.de`.

Attachments:

* [linux64-mc-slip.tar.gz](https://github.com/agra-uni-bremen/riscv-vp/files/8339992/linux64-mc-slip.tar.gz) (takes a while to boot since it uses multicore linux-vp and SystemC doesn't parallelize the different cores, login is root:agra)
